File: web-client/src/api/session.ts
import {
  type QueryClient,
  queryOptions,
  useMutation,
  useQuery,
  useQueryClient,
} from '@tanstack/react-query'
import { z } from 'zod'
import { ApiError, api, hasCsrfCookie, restoreCsrfCompanion, unwrap } from './client'
import { handleIdentityChange } from './identity-change'
import { announceIdentityChange, forgetSessionEnd, readEndedSession, rememberSessionEnd, synchronizeSessionEnd } from './browser-session'
import { closeRealtimeConnections } from './realtime/connection'
import { clearAppEntered } from '@/lib/landing-redirect'
import type { components } from './schema'

export type Session = components['schemas']['SessionResponse']
export type SessionUser = components['schemas']['SessionUser']

export const SESSION_QUERY_KEY = ['session'] as const

// TanStack Query only single-flights `/v1/session` within one tab's
// QueryClient. Several cold tabs opened at once each have their own
// QueryClient, so each fires its own request before the browser has a
// `session` cookie, and each mints its own guest — the last `Set-Cookie` wins
// and the other tabs are left holding a stale identity (#824). This lock
// widens the singleflight to the whole origin for the one race that matters:
// the cold bootstrap. Once a `csrf_token` cookie is readable, a session
// already exists, so later calls skip the lock entirely.
const SESSION_LOCK_NAME = 'fortymm:session-bootstrap'
const SESSION_LOCK_STORAGE_KEY = 'fortymm:session-bootstrap:lock'
const SESSION_LOCK_TTL_MS = 10_000
const SESSION_LOCK_POLL_MS = 50

// The lock record is persisted client state (localStorage), so it's untrusted
// on read — a different app version, a manual edit, or a truncated write could
// leave anything there. Parse it at the boundary instead of casting the
// `JSON.parse` result (see `.claude/rules/parse-at-boundaries.md`); a shape
// mismatch is treated as "no lock", exactly like a parse/JSON error.
const storageLockSchema = z.object({
  owner: z.string(),
  expires: z.number(),
})
type StorageLockRecord = z.infer<typeof storageLockSchema>

export function readStorageLock(): StorageLockRecord | null {
  const raw = localStorage.getItem(SESSION_LOCK_STORAGE_KEY)
  if (!raw) return null
  try {
    const parsed = storageLockSchema.safeParse(JSON.parse(raw))
    return parsed.success ? parsed.data : null
  } catch {
    return null
  }
}

// Fallback single-flight for browsers without the Web Locks API. localStorage
// writes aren't atomic across tabs, so this can't guarantee mutual exclusion
// the way `navigator.locks` does — but it narrows the race window to
// milliseconds, and the TTL means a tab that crashes mid-bootstrap can never
// wedge the others (they just wait out the TTL and proceed).
async function withStorageLock<T>(fn: () => Promise<T>, alwaysLock = false, requireStorage = alwaysLock): Promise<T> {
  const owner = `${Date.now()}-${Math.random()}`
  const deadline = Date.now() + SESSION_LOCK_TTL_MS
  for (;;) {
    try {
      const held = readStorageLock()
      if (!held || held.expires < Date.now()) {
        localStorage.setItem(
          SESSION_LOCK_STORAGE_KEY,
          JSON.stringify({ owner, expires: Date.now() + SESSION_LOCK_TTL_MS }),
        )
        // Yield a tick, then re-read: narrows (but can't eliminate) the window
        // where two tabs both saw no lock and both wrote.
        await new Promise((resolve) => setTimeout(resolve, 0))
        if (readStorageLock()?.owner === owner) break
      }
    } catch {
      // Destructive recovery needs cross-tab exclusion; an in-tab queue
      // cannot safely replace it when the shared backend is unavailable.
      if (requireStorage) throw new Error('Session recovery is unavailable. Please enable browser storage and try again.')
      return fn()
    }
    if (Date.now() > deadline) throw new Error('Another tab is updating your session. Please try again.')
    if (!alwaysLock && hasCsrfCookie()) return fn()
    await new Promise((resolve) => setTimeout(resolve, SESSION_LOCK_POLL_MS))
  }
  // Renew while the request is pending: a slow network response must not
  // turn an active recovery into an expired lock that another tab can take.
  const renewal = setInterval(() => {
    try {
      if (readStorageLock()?.owner === owner) {
        localStorage.setItem(SESSION_LOCK_STORAGE_KEY,
          JSON.stringify({ owner, expires: Date.now() + SESSION_LOCK_TTL_MS }))
      }
    } catch { clearInterval(renewal) }
  }, SESSION_LOCK_TTL_MS / 3)
  try {
    return await fn()
  } finally {
    clearInterval(renewal)
    try {
      if (readStorageLock()?.owner === owner) localStorage.removeItem(SESSION_LOCK_STORAGE_KEY)
    } catch { /* Storage became unavailable during the request. */ }
  }
}

let localSessionQueue: Promise<unknown> = Promise.resolve()
function withLocalSessionLock<T>(fn: () => Promise<T>): Promise<T> {
  const result = localSessionQueue.then(fn, fn)
  localSessionQueue = result.catch(() => undefined)
  return result
}

/** Single-flights the `/v1/session` cold-bootstrap request across every tab
 * on the origin, not just within one TanStack QueryClient. */
async function withSessionBootstrapLock<T>(fn: () => Promise<T>, mode: 'bootstrap' | 'recovery' | 'logout' = 'bootstrap'): Promise<T> {
  const alwaysLock = mode !== 'bootstrap'
  // A session already exists — no mint race to guard against.
  if (!alwaysLock && hasCsrfCookie()) return fn()
  const run = () => {
    if (mode === 'recovery') synchronizeSessionEnd()
    if (mode === 'logout') {
      // Refresh stale retry state when possible, but storage failures must
      // never prevent revocation of the current server-side credential.
      try { synchronizeSessionEnd() } catch { /* Revocation remains available. */ }
    }
    return fn()
  }
  if (typeof navigator !== 'undefined' && navigator.locks?.request) {
    return navigator.locks.request(SESSION_LOCK_NAME, run)
  }
  return withLocalSessionLock(() => withStorageLock(run, alwaysLock, mode === 'recovery'))
}

export function sessionQueryOptions() {
  return queryOptions({
    queryKey: SESSION_QUERY_KEY,
    queryFn: (): Promise<Session> =>
      withSessionBootstrapLock(async () => {
        const ended = readEndedSession()
        if (ended) {
          throw new ApiError(401, ended.message, 'load session', {
            detail: { code: 'session_ended', ...ended },
          })
        }
        return unwrap('load session', await api.GET('/v1/session'))
      }),
    staleTime: 1000 * 60 * 5,
    // Don't retry a 401 (session merged away): the 401 already cleared the
    // cookie, so a retry would silently mint a *new* guest and race the
    // redirect-to-login. Transient errors still retry.
    retry: (failureCount, error) =>
      !(error instanceof ApiError && error.status === 401) && failureCount < 3,
  })
}

export function useSession() {
  return useQuery(sessionQueryOptions())
}

export type EmailStatus = 'guest' | 'pending' | 'verified'

export function deriveEmailStatus({
  email,
  confirmedAt,
  pendingEmail,
}: {
  email: string | null
  confirmedAt: string | null
  pendingEmail: string | null
}): EmailStatus {
  // A pending change wins even when the user is already verified — the FE
  // still has something to nudge them to complete.
  if (pendingEmail) return 'pending'
  if (confirmedAt) return 'verified'
  if (email) return 'pending'
  return 'guest'
}

/** True when the current session carries `name` in its permissions list. */
export function useHasPermission(name: string): boolean {
  const { data } = useSession()
  return data?.data.user.permissions.includes(name) ?? false
}

// Callers that want inline 4xx error handling (e.g. ChangeUsernameDialog)
// must await this via mutateAsync and catch ApiError themselves.
export function useUpdateUsername() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async (username: string): Promise<Session> =>
      unwrap(
        'update username',
        await api.PATCH('/v1/me', { body: { username } }),
      ),
    onSuccess: (session) => {
      qc.setQueryData(SESSION_QUERY_KEY, session)
    },
  })
}

export interface SetEmailInput {
  email: string
  captchaToken: string
  // Honeypot — the form must include this field on the wire even when
  // empty so the server's bot-check semantics match the FE form shape.
  honeypot?: string
}

export function useSetEmail() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({
      email,
      captchaToken,
      honeypot = '',
    }: SetEmailInput): Promise<Session> =>
      unwrap(
        'set email',
        await api.POST('/v1/me/email', {
          body: {
            email,
            captcha_token: captchaToken,
            fmm_hp_token: honeypot,
          },
        }),
      ),
    onSuccess: (session) => {
      qc.setQueryData(SESSION_QUERY_KEY, session)
    },
  })
}

export function useResendEmailConfirmation() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({
      captchaToken,
      honeypot = '',
    }: {
      captchaToken: string
      honeypot?: string
    }): Promise<Session> =>
      unwrap(
        'resend confirmation',
        await api.POST('/v1/me/email/resend', {
          body: { captcha_token: captchaToken, fmm_hp_token: honeypot },
        }),
      ),
    onSuccess: (session) => {
      qc.setQueryData(SESSION_QUERY_KEY, session)
    },
  })
}

export type MergePreview = components['schemas']['MergePreview']

const accountSwitchSchema = z.object({
  from_user_id: z.string(),
  from_username: z.string(),
  to_username: z.string(),
})
const mergePreviewSchema = z.object({
  is_merge: z.boolean(),
  owner_username: z.string().nullable().optional(),
  guest_username: z.string().nullable().optional(),
  guest_matches_count: z.number().int().nonnegative().default(0),
  adopts_guest_username: z.boolean().default(false),
  account_switch: accountSwitchSchema.nullable().optional(),
})

/** Side-effect-free look at an emailed link, to decide whether to show the
 * "bring N matches over?" gate before finalizing. */
export function useMergePreview() {
  return useMutation({
    mutationFn: async (token: string): Promise<MergePreview> =>
      mergePreviewSchema.parse(
        unwrap('check link', await api.POST('/v1/merge/preview', { body: { token } })),
      ),
  })
}

/** Input for the finalize mutations. `skipMerge` is the gate's "not now": sign
 * in without folding the guest's matches in. */
export interface FinalizeTokenInput {
  token: string
  skipMerge?: boolean
  switchFromUserId?: string
}

/** Seed `SESSION_QUERY_KEY` from a sign-in/confirm response. `GET /v1/session`
 * never returns `merged` — strip it before caching so a future
 * `useSession().data.merged` read can't see this mutation's stale value for
 * the full 5-minute staleTime (#239). */
function cacheSession(qc: QueryClient, session: Session): void {
  qc.setQueryData(SESSION_QUERY_KEY, { ...session, merged: null })
}

async function afterPendingLogout<T>(fn: () => Promise<T>): Promise<T> {
  if (!readEndedSession()?.logoutPending) return fn()
  return withSessionBootstrapLock(async () => {
    if (readEndedSession()?.logoutPending) {
      restoreCsrfCompanion()
      unwrap('finish sign out', await api.DELETE('/v1/session'), { allowEmpty: true })
      rememberSessionEnd({ message: 'You have signed out. Sign in to continue.', logoutPending: false }, { notifyLocal: false })
    }
    return fn()
  }, 'recovery')
}

export function useConfirmEmail() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({
      token,
      skipMerge = false,
      switchFromUserId,
    }: FinalizeTokenInput): Promise<Session> =>
      afterPendingLogout(async () => unwrap(
        'confirm email',
        await api.POST('/v1/me/email/confirm', {
          body: { token, skip_merge: skipMerge, switch_from_user_id: switchFromUserId },
        }),
      )),
    // This can sign the caller into a *different* existing account
    // (skip_merge). Drop the prior identity's cached per-user data first — the
    // same leak useLogout guards against (#754) — then reseed the session so
    // it doesn't need a refetch. `clearAppEntered` is deliberately NOT part of
    // it: this browser is arriving, not leaving.
    onSuccess: (session) => {
      handleIdentityChange({
        closeRealtime: closeRealtimeConnections,
        clearQueryCache: () => qc.clear(),
      })
      cacheSession(qc, session)
      forgetSessionEnd()
      announceIdentityChange()
    },
  })
}

export interface RequestLoginInput {
  email: string
  captchaToken: string
  honeypot?: string
}

export function useRequestLogin() {
  return useMutation({
    mutationFn: async ({
      email,
      captchaToken,
      honeypot = '',
    }: RequestLoginInput) =>
      unwrap(
        'request sign-in link',
        await api.POST('/v1/login/request', {
          body: {
            email,
            captcha_token: captchaToken,
            fmm_hp_token: honeypot,
          },
        }),
      ),
  })
}

export function useLogout(retryOnly = false) {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (): Promise<void> => withSessionBootstrapLock(async () => {
      if (retryOnly && !readEndedSession()?.logoutPending) return
      restoreCsrfCompanion()
      rememberSessionEnd({ message: 'Sign-out is not complete. Retry to finish signing out.', logoutPending: true })
      unwrap('sign out', await api.DELETE('/v1/session'), { allowEmpty: true })
      if (readEndedSession()?.logoutPending) {
        rememberSessionEnd({ message: 'You have signed out. Sign in to continue.', logoutPending: false })
      }
    }, 'logout'),
    // Drop ALL cached per-user data (matches, dashboard, players, ...), not
    // just SESSION_QUERY_KEY — otherwise the prior user's BFF responses leak
    // into the next ephemeral session.
    //
    // Through the shared sequence, not inline: this is the path MOST exposed to
    // the ordering hazard, because it always fires from inside `_app` with a
    // live `/v1/stream` open (the user menu and the settings footer). The
    // caller owns the navigation that follows, so no `navigateToLogin` here.
    onSuccess: () => {
      if (readEndedSession()?.logoutPending !== false) return
      handleIdentityChange({
        closeRealtime: closeRealtimeConnections,
        clearAppEntered,
        clearQueryCache: () => qc.clear(),
      })
    },
  })
}

export function useConsumeLoginToken() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: async ({
      token,
      skipMerge = false,
      switchFromUserId,
    }: FinalizeTokenInput): Promise<Session> =>
      afterPendingLogout(async () => unwrap(
        'sign in',
        await api.POST('/v1/login/consume', {
          body: { token, skip_merge: skipMerge, switch_from_user_id: switchFromUserId },
        }),
      )),
    // Same identity-leak guard as useConfirmEmail (#754): this can sign a
    // browsing guest into a different existing account.
    onSuccess: (session) => {
      handleIdentityChange({
        closeRealtime: closeRealtimeConnections,
        clearQueryCache: () => qc.clear(),
      })
      cacheSession(qc, session)
      forgetSessionEnd()
      announceIdentityChange()
    },
  })
}

// The bare address auth mail really sends from (`GET /v1/login/sender`), for
// the `/login/sent` receipt row (#1466 defect 1). Static, deployment-wide,
// takes no input and reads no cookie — so unlike `useSession` it's safe to
// call from a bookmarked `/login/sent` without minting a guest.
export const LOGIN_SENDER_QUERY_KEY = ['login-sender'] as const

const loginSenderResponseSchema = z.object({ address: z.string().nullable() })

export function loginSenderQueryOptions() {
  return queryOptions({
    queryKey: LOGIN_SENDER_QUERY_KEY,
    // The Zod parse is still the runtime boundary guarantee
    // (`.claude/rules/parse-at-boundaries.md`) even though `api.GET` now
    // types this path from the regenerated `schema.d.ts` — a type
    // annotation is a compile-time claim, not a runtime check.
    queryFn: async (): Promise<string | null> => {
      const body = unwrap('load sign-in sender', await api.GET('/v1/login/sender'))
      return loginSenderResponseSchema.parse(body).address
    },
    // A single constant for the whole deployment (Settings.email_from) — it
    // cannot change mid-session, so there's nothing to ever refetch for.
    staleTime: Infinity,
    retry: false,
  })
}

/** The bare sender address, or `undefined` while loading/on failure — callers
 * must render fine either way (see `EmailReceipt`'s optional `sender` prop):
 * this is receipt trivia, never something worth blocking `/login/sent` on. */
export function useLoginSender() {
  return useQuery(loginSenderQueryOptions())
}

/** An explicit choice to abandon the ended session and start a separate guest. */
export function useStartNewGuest() {
  const qc = useQueryClient()
  return useMutation({
    mutationFn: (): Promise<Session> => withSessionBootstrapLock(async () => {
      // A preceding recovery in another tab may already have replaced the
      // ended session. Reuse that identity instead of deleting it again.
      if (readEndedSession()) {
        restoreCsrfCompanion()
        unwrap('clear old session', await api.DELETE('/v1/session'), { allowEmpty: true })
      }
      const session = unwrap('start a new guest', await api.GET('/v1/session'))
      // Publish completion before releasing the origin-wide lock.
      forgetSessionEnd()
      announceIdentityChange()
      return session
    }, 'recovery'),
    onSuccess: (session) => {
      handleIdentityChange({
        closeRealtime: closeRealtimeConnections,
        clearQueryCache: () => qc.clear(),
      })
      cacheSession(qc, session)
    },
  })
}

const switchConflictSchema = z.object({
  detail: z.object({
    code: z.literal('account_switch_required'),
    account_switch: accountSwitchSchema.nullable(),
  }),
})

export function accountSwitchConflict(error: unknown) {
  if (!(error instanceof ApiError) || error.status !== 409) return null
  const result = switchConflictSchema.safeParse(error.body)
  return result.success ? result.data.detail : null
}

File: api/app/schemas/session.py
import uuid
from datetime import datetime
from typing import Annotated

from pydantic import AfterValidator, BaseModel, ConfigDict, EmailStr, Field, RootModel

# RFC 5321 §4.5.3.1.1 caps the local part (before the ``@``) at 64 octets and
# the whole address at 254. ``EmailStr`` / email-validator enforces the 254
# total but NOT the 64-char local part, so an oversize local part validates and
# we'd enqueue mail to it (#615). Reject it explicitly on inbound addresses.
EMAIL_LOCAL_PART_MAX_LENGTH = 64


def _validate_local_part_length(value: str) -> str:
    local_part, _, _ = value.rpartition("@")
    if len(local_part) > EMAIL_LOCAL_PART_MAX_LENGTH:
        raise ValueError(
            "The email address's local part is too long "
            f"(maximum {EMAIL_LOCAL_PART_MAX_LENGTH} characters)."
        )
    return value


# Use for inbound addresses we'll act on (send mail / persist). Response echoes
# can stay plain ``EmailStr`` — they re-emit an already-validated value.
BoundedEmailStr = Annotated[EmailStr, AfterValidator(_validate_local_part_length)]

# Lowercase alphanumerics with optional dots/hyphens/underscores between them.
# Must start and end with an alphanumeric so we don't store names that look
# like punctuation. 40 chars admits the auto-generated `slug-slug-XXXXXXXX`
# format with headroom while keeping display sane.
USERNAME_PATTERN = r"^[a-z0-9](?:[a-z0-9._-]{1,38}[a-z0-9])?$"
USERNAME_MIN_LENGTH = 3
USERNAME_MAX_LENGTH = 40


class SessionUser(BaseModel):
    # The caller's own user id. Returned only to the holder of the session
    # cookie, and the only place the API tells anyone who they are: without it
    # clients have to guess at "is this me?" by comparing usernames, which a
    # rename breaks. Never null — a guest minted by ``GET /v1/session`` is a
    # real user row with a real id.
    id: uuid.UUID
    username: str
    permissions: list[str]
    email: str | None = None
    confirmed_at: datetime | None = None
    pending_email: str | None = None


class SessionData(BaseModel):
    user: SessionUser


class MergeSummary(BaseModel):
    """Reported by sign-in / email-confirm responses when the call merged the
    caller's ephemeral session into a different (verified) account. Drives the
    "we brought your N matches with you" toast."""

    matches_moved: int


class SessionResponse(BaseModel):
    data: SessionData
    merged: MergeSummary | None = None


class UpdateCurrentUserRequest(BaseModel):
    username: str = Field(
        min_length=USERNAME_MIN_LENGTH,
        max_length=USERNAME_MAX_LENGTH,
        pattern=USERNAME_PATTERN,
    )


class CaptchaProtectedRequest(BaseModel):
    """Base for endpoints gated by Cloudflare Turnstile + an off-screen
    honeypot. ``fmm_hp_token`` is the honeypot — bots fill every field;
    humans never see it. Endpoints short-circuit as if successful when it's
    filled, so the bot doesn't learn the field is a trap.

    The field name deliberately avoids canonical identity-profile names
    (``website``, ``address``, ``phone``) because password managers and
    browser autofillers ignore ``autocomplete=off`` for those and would
    splash a real user's saved value into the trap."""

    captcha_token: str = Field(min_length=1, max_length=4096)
    fmm_hp_token: str = Field(default="", max_length=512)


class SetEmailRequest(CaptchaProtectedRequest):
    email: BoundedEmailStr


class ResendEmailRequest(CaptchaProtectedRequest):
    pass


class ConfirmEmailRequest(BaseModel):
    switch_from_user_id: uuid.UUID | None = None
    token: str = Field(min_length=1, max_length=512)
    # When the link would fold a guest into this account, the client can offer
    # the owner a "sign in but don't bring those matches" choice. Defaults to
    # merging (the common, desired path).
    skip_merge: bool = False


class ConfirmEmailErrorDetail(BaseModel):
    """The coded detail a ``400`` from ``/v1/me/email/confirm`` carries for one
    specific failure — a confirmation link a newer resend superseded
    (#1616). ``code`` is the machine-readable reason clients branch on
    (``replaced``); ``message`` is the server's own sentence. It is one of the
    two ``400`` bodies that endpoint can return (see
    ``ConfirmEmail400Response``); every other dead link carries the
    plain-string detail of ``PlainDetailErrorResponse``."""

    code: str
    message: str


class ConfirmEmailErrorResponse(BaseModel):
    """The coded 400 body ``confirm_email`` raises for a superseded
    confirmation link — ``{"detail": {"code": ..., "message": ...}}`` (#1616).
    One of the two ``400`` bodies that endpoint can return (see
    ``ConfirmEmail400Response``)."""

    detail: ConfirmEmailErrorDetail


class PlainDetailErrorResponse(BaseModel):
    """The default FastAPI error body — ``{"detail": "<sentence>"}``, what
    ``HTTPException(detail=str)`` produces. The 400 ``confirm_email`` returns
    for every dead confirmation link except the superseded one: invalid,
    expired, or a replaced row whose newer link is itself dead. Alongside
    ``ConfirmEmailErrorResponse`` it makes up ``ConfirmEmail400Response``."""

    detail: str


class ConfirmEmail400Response(
    RootModel[ConfirmEmailErrorResponse | PlainDetailErrorResponse]
):
    """Both 400 bodies ``/v1/me/email/confirm`` can return. A link a newer
    resend replaced carries the coded ``ConfirmEmailErrorResponse`` shape
    (#1616); every other dead link — invalid, expired, or a replaced row whose
    newer link is itself gone — carries the plain-string
    ``PlainDetailErrorResponse`` shape. Declared on the route's ``responses=``
    as this union, because a generated client decoding every 400 as only the
    coded shape would fail on a normal rejected link before it could handle
    it (#1632)."""

    root: ConfirmEmailErrorResponse | PlainDetailErrorResponse


class RequestLoginRequest(CaptchaProtectedRequest):
    email: BoundedEmailStr


class LoginRequestAccepted(BaseModel):
    """202 body for the magic-link request endpoint. Always echoes the
    submitted address — identical whether or not it maps to a real account,
    so the response leaks nothing about which addresses are registered."""

    email: EmailStr


class LoginSenderResponse(BaseModel):
    """Body for ``GET /v1/login/sender`` — the bare address auth mail really
    sends from, parsed out of ``Settings.email_from``'s RFC 5322 display form
    (``FortyMM <noreply@fortymm.com>`` -> ``noreply@fortymm.com``). A static,
    deployment-wide constant, not user- or request-specific, so it is safe to
    serve with no cookie and no captcha. ``None`` only when the configured
    value doesn't parse to an address at all — the client renders without a
    sender row rather than a broken one."""

    address: str | None


class ConsumeLoginRequest(BaseModel):
    switch_from_user_id: uuid.UUID | None = None
    token: str = Field(min_length=1, max_length=512)
    # See ConfirmEmailRequest.skip_merge — sign in without folding the recorded
    # guest's matches in.
    skip_merge: bool = False


class MergePreviewRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    token: str = Field(min_length=1, max_length=512)


class AccountSwitchPreview(BaseModel):
    from_user_id: uuid.UUID
    from_username: str
    to_username: str


class MergePreview(BaseModel):
    """Side-effect-free look at an emailed link before it's consumed, so the
    client can decide whether to show a "bring N matches over?" confirmation.

    ``is_merge`` is true only for a link that would fold a guest into another
    account (a settings merge token, or a sign-in token that recorded a
    requesting guest). The client shows the gate only when there are matches to
    carry (``guest_matches_count > 0``); otherwise it finalizes silently.

    ``adopts_guest_username`` is true only when the link is a *first* sign-in,
    where the account being signed into was minted moments ago and its
    ``owner_username`` is a throwaway generated slug. Accepting the merge moves
    ``guest_username`` onto it; declining leaves the generated one. On every
    other merge the username does not move, so the gate must not promise it
    will."""

    account_switch: AccountSwitchPreview | None = None
    is_merge: bool
    owner_username: str | None = None
    guest_username: str | None = None
    guest_matches_count: int = 0
    adopts_guest_username: bool = False
